### PR TITLE
SONARJAVA-2374 fix FPs on exceptional paths and constructor args

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 1]
```diff
@@ -45,6 +45,7 @@
 import org.sonar.java.se.checks.OptionalGetBeforeIsPresentCheck;
 import org.sonar.java.se.checks.SECheck;
 import org.sonar.java.se.checks.StreamConsumedCheck;
+import org.sonar.java.se.checks.StreamNotConsumedCheck;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: MINOR Remove this unused import 'org.sonar.java.se.checks.StreamNotConsumedCheck'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ExplodedGraph.java#L49](https://github.com/SonarSource/sonar-java/blob/36ed42f15c255cb2d9855a2d981e59b42ee8850d/java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java#L49): Replace this "Map.get()" and condition with a call to "Map.computeIfAbsent()". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3824)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ExplodedGraph.java#L49](https://github.com/SonarSource/sonar-java/blob/a9f303a169f78032aeaa2407a37d2abd4f42a1dd/java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java#L49): Replace this "Map.get()" and condition with a call to "Map.computeIfAbsent()". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3824)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/checks/StreamConsumedCheck.java [hunk 2]
```diff
@@ -39,10 +39,14 @@
 import org.sonar.plugins.java.api.tree.IdentifierTree;
 import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
 import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
 import org.sonar.plugins.java.api.tree.ReturnStatementTree;
 import org.sonar.plugins.java.api.tree.Tree;
 
+import java.util.ArrayDeque;
 import java.util.Collections;
+import java.util.Deque;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/StreamConsumedCheck.java]: MINOR Remove this unused import 'java.util.Deque'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-frontend/src/main/java/org/sonar/java/se/checks/StreamConsumedCheck.java [hunk 3]
```diff
@@ -39,10 +39,14 @@
...
+import java.util.HashSet;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/StreamConsumedCheck.java]: MINOR Remove this unused import 'java.util.HashSet'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-frontend/src/main/java/org/sonar/java/se/checks/StreamConsumedCheck.java [hunk 4]
```diff
@@ -39,10 +39,14 @@
 import org.sonar.plugins.java.api.tree.IdentifierTree;
 import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
 import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
 import org.sonar.plugins.java.api.tree.ReturnStatementTree;
 import org.sonar.plugins.java.api.tree.Tree;
 
+import java.util.ArrayDeque;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/StreamConsumedCheck.java]: MINOR Remove this unused import 'java.util.ArrayDeque'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: java-frontend/src/main/java/org/sonar/java/se/checks/StreamConsumedCheck.java [hunk 5]
```diff
@@ -75,7 +76,22 @@ public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
...
+      return removeConstraintOnArgs(context.getState(), syntaxNode);
+    }
+    ProgramState state = context.getState();
+    if (state.peekValue() instanceof SymbolicValue.ExceptionalSymbolicValue) {
+      state = removeConstraintOnExceptionalPath(context);
+    }
+    return state;
+  }
+
+  private static ProgramState removeConstraintOnExceptionalPath(CheckerContext context) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/StreamConsumedCheck.java]: matter of naming : this method does not need to know it is working on Exceptional Path. I would go for : `ProgramState removeNotConsumedConstraint(ProgramState state)`

File: java-frontend/src/main/java/org/sonar/java/se/checks/StreamConsumedCheck.java [hunk 6]
```diff
@@ -75,7 +76,22 @@ public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
     if (syntaxNode.is(Tree.Kind.METHOD_INVOCATION)) {
       return handleMethodInvocation(context, syntaxNode);
     }
-    return context.getState();
+    if (syntaxNode.is(Tree.Kind.NEW_CLASS)) {
+      return removeConstraintOnArgs(context.getState(), syntaxNode);
+    }
+    ProgramState state = context.getState();
+    if (state.peekValue() instanceof SymbolicValue.ExceptionalSymbolicValue) {
+      state = removeConstraintOnExceptionalPath(context);
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/StreamConsumedCheck.java]: As stated in next comment, you only need to pass the state, not the context.

File: java-frontend/src/test/files/se/StreamNotConsumedCheck.java [hunk 7]
```diff
@@ -63,4 +63,55 @@ void testField() {
...
+
+      Stream.of(1,2,3).skip(skipCount()); // Noncompliant
+    } catch (Exception ex) {
+      System.out.println("Exception!");
+    }
+  }
+
+  void exceptionalPathFN(Stream<Object> stream) {
+    try {
+      stream = stream.skip(skipCount());  // Compliant this pipeline is not consumed on exceptional path
```
Comment (benzonico) [inline java-frontend/src/test/files/se/StreamNotConsumedCheck.java]: comment is confusing with the method name : is that a false negative ? if yes, state it in the comment.

File: java-frontend/src/test/files/se/StreamNotConsumedCheck.java [hunk 8]
```diff
@@ -63,4 +63,55 @@ void testField() {
...
+    } catch (Exception ex) {
+      return;
+    }
+    stream.forEach(System.out::println);
+  }
+
+  void testss(Stream<Object> stream) {
+    stream.filter(e -> true);  // FN because SV is reused
```
Comment (benzonico) [inline java-frontend/src/test/files/se/StreamNotConsumedCheck.java]: I challenge you to understand that comment 6 months from now : please make it clear that as filter is returning the same SV as evaluation of stream, it is consumed by the second line whereas it is not the case in reality.

**Pull request merged**